Reset Controller Driver

andrei-edward-popa opened the pull request:
This is my initial work and approach for a generic reset controller driver that can be used in Zephyr. I am open to any ideas for improvement of this driver, but this was my first thought.

carlocaione commented on drivers/reset/reset.c:
```diff
@@ -0,0 +1,149 @@
+struct reset_generic_config {
+	uint8_t reg_width;
+	uint8_t active_low;
+	uint32_t base_address;
```
`uintptr_t`

carlocaione commented on drivers/reset/reset.c:
```diff
@@ -0,0 +1,149 @@
+		value &= ~BIT(regbit);
+	}
+
+	ret = reset_generic_write_register(dev, offset, value);
```
just `return reset_...`

carlocaione commented on drivers/reset/reset.c:
```diff
@@ -0,0 +1,149 @@
+	static const struct reset_generic_config reset_generic_config_##idx = {		\
+		.reg_width = DT_INST_PROP_OR(idx, reg_width, 4),			\
+		.active_low = DT_INST_PROP_OR(idx, active_low, 0),			\
+		.base_address = DT_INST_REG_ADDR_BY_IDX(idx, 0),			\
```
why not simply `DT_INST_REG_ADDR()`?

carlocaione commented on drivers/reset/reset.c:
```diff
@@ -0,0 +1,149 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT reset
```
do you have any hardware implementing this? This looks like a reset controller leveraging some syscon area.

andrei-edward-popa replied:
> do you have any hardware implementing this? This looks like a reset controller leveraging some syscon area.

Yes, I have an implementation for Raspberry Pi Pico, I will update the code soon.

carlocaione commented:
two more points:
- Do you have any hardware using a reset controller? It would be great to have at least a sample or test or hardware using this API
- Please use the `device_mmio` framework (see https://github.com/zephyrproject-rtos/zephyr/blob/main/include/sys/device_mmio.h and https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/syscon/syscon.c as example)

andrei-edward-popa commented on drivers/reset/reset.c:
```diff
@@ -0,0 +1,149 @@
+/*
```
No, you are right, I will update it with a hardware-specific implementation for rpi_pico soon.

yonsch replied:
Not sure I understood this correctly, but is this API intended to retrieve the reset cause? If it is, how is it different from hwinfo?
Regarding the rpi pico, there's a hwinfo PR here: https://github.com/zephyrproject-rtos/zephyr/pull/42558

andrei-edward-popa replied:
If I understand correctly what you did in hwinfo, you talk about the RP2040 microcontroller reset cause, right? This reset controller driver is intended to reset individual devices like UART, I2C etc.

erwango commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -302,6 +311,9 @@ static const struct uart_driver_api uart_rpi_driver_api = {
 		.uart_dev = uart##idx,						\
 		.uart_regs = (uart_hw_t *)uart##idx,				\
 		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),			\
+		.reset_ctrl = DEVICE_DT_GET(DT_INST_PHANDLE(idx, resets)),	\
+		.reset_offset = DT_INST_PHA_BY_IDX(idx, resets, 0, offset),	\
+		.reset_bit = DT_INST_PHA_BY_IDX(idx, resets, 0, regbit),	\
```
A single concise macro, similar to what `PINCTRL_DT_INST_DEV_CONFIG_GET` is doing would be nice IMHO

andrei-edward-popa replied:
Okay, I will try to create some macros for this.

andrei-edward-popa replied:
> A single concise macro, similar to what `PINCTRL_DT_INST_DEV_CONFIG_GET` is doing would be nice IMHO

I did something, probably not perfect, but if you have another suggestions I will gladly receive them.

carlocaione commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,179 @@
+ * @brief Reset Controller driver API
+ */
+__subsystem struct reset_driver_api {
+	reset_api_status status;
+	reset_api_assert assert;
+	reset_api_deassert deassert;
+	reset_api_reset reset;
```
I'm a bit confused about the usage of `offset` and `regbit`. You are implicitly assuming that the reset controller is using some MMIO mapping (offset + regbit) but this could not be the case for every hardware.

For example Linux just uses a generic `unsigned long id` for the line.

carlocaione commented on drivers/reset/Kconfig:
```diff
@@ -0,0 +1,32 @@
+menuconfig RESET
+	bool "Reset Controller drivers"
+	help
+	  Reset Controller drivers. Reset controller node represents
+	  a region containing the base address of a reset controller device.
+	  The typical use-case is for some other node's drivers to acquire a
+	  reference to the reset controller node together with some reset
+	  information such as offset to reset register and reset bit and
+	  use syscalls to call the reset controller API functions.
```
Indeed what you are describing here is a very special kind of reset controller. I understand that this probably is  99% of the reset controllers out there, but I think we should be more generic.

carlocaione commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -302,6 +311,9 @@ static const struct uart_driver_api uart_rpi_driver_api = {
 		.uart_dev = uart##idx,						\
 		.uart_regs = (uart_hw_t *)uart##idx,				\
 		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),			\
+		.reset_ctrl = DEVICE_DT_GET(DT_INST_PHANDLE(idx, resets)),	\
+		.reset_offset = DT_INST_PHA_BY_IDX(idx, resets, 0, offset),	\
+		.reset_bit = DT_INST_PHA_BY_IDX(idx, resets, 0, regbit),	\
```
Are these being used?

andrei-edward-popa replied:
The reset controller is not used right now, it will be used at the moment of a working implementation of a clock controller for pico to replace uart_init() pico sdk function with an implementation independent of their sdk.

carlocaione commented on dts/bindings/reset/reset-controller.yaml:
```diff
@@ -0,0 +1,22 @@
+
+description: Reset Controller
+
+include: base.yaml
```
This should be included by the yaml file of the reset controllers, not by this file.

carlocaione commented on dts/bindings/reset/reset-controller.yaml:
```diff
@@ -0,0 +1,22 @@
+include: base.yaml
+
+properties:
+    reg:
+      required: true
```
Not needed here.

carlocaione commented on dts/bindings/reset/reset-controller.yaml:
```diff
@@ -0,0 +1,22 @@
+properties:
+    reg:
+      required: true
+    reg-width:
+      type: int
+      required: false
+      description: The width of the reset registers in bytes. Default is 4 bytes.
+    active-low:
+      type: int
+      required: false
+      description: Set if reset is active low. Default is 0, which means active-high.
```
This is not something related to the reset API. This should be moved to the reset controller yaml file.

carlocaione commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,192 @@
+#endif
+
+/** Reset controller device configuration. */
+struct reset_dev_config {
+	/** Reset controller device. */
+	const struct device *dev;
+	/** Reset line. */
+	uint32_t id;
+};
+
+/**
+ * @brief Helper macro to initialize reset controller device config.
+ *
+ * @param inst Instance number.
+ * @param id Reset line.
+ */
+#define RESET_DEV_CONFIG_INIT(inst, id)						\
+	{									\
+		.dev = DEVICE_DT_GET(DT_INST_RESET_CTLR(inst)),			\
+		.id = DT_INST_RESET_CELL(inst, id),				\
+	}
```
I'm a bit puzzled about this. Who is going to use the `struct reset_dev_config` and how? Take for example https://github.com/zephyrproject-rtos/zephyr/blob/8a8bc392dc93a78ec3a84ba68811b46a3cc73ffa/include/drivers/mbox.h#L92-L98
In this case we have added some helper functions (see the `mbox_*` functions) that leverage that struct. But in your case how you are supposed to use this struct and why is it there if there is no user?

andrei-edward-popa replied:
I tried to create this struct for getting the reset controller device and its line into some device driver (you can see an example of this in drivers/serial/uart_rpi_pico.c) and if you want to reset the device you can say something like this:
reset_reset(reset->dev, reset->line);
In the line below, 'reset' is a pointer to a reset_dev_config.
I'm not sure how to do it in another way, but if you have any idea I will gladly receive it.

carlocaione replied:
> if you want to reset the device you can say something like this:
> reset_reset(reset->dev, reset->line);

yes, but that makes sense if you have a wrapper at API level that can unpack that for you, for example `reset((struct reset_dev_config *) conf)`. In your case that wrapper is not being defined at API level so this struct is simply unused (I mean, the same struct can be created by the driver if it is going to be used right?)

andrei-edward-popa replied:
Okay, yes, I think I understand it, I will try to create some wrappers.

andrei-edward-popa replied:
@carlocaione I have already a syscall called reset_reset as you saw, and this takes a dev pointer and an id and now I want to create another syscall that takes a reset_dev_config. How can I call it (I'm talking about its name)?

carlocaione replied:
no, do not create more syscalls. All your syscalls are already taking a dev pointer and an id, why don't you simply change the syscalls you have already to take a `struct reset_dev_config` instead?

andrei-edward-popa replied:
Okay, that's cool.

ceolin replied:
you should be using `device *` and not this struct. The reason is that `struct device` is a kernel object and because of this you can grant or revoke access per thread. You can also go additional validations like if the type of the device is right ...

andrei-edward-popa replied:
@ceolin What about the MBOX device? It is using a structure like I did. Maybe @carlocaione or @gmarull can clarify this? I'm not sure how to do it in another way and some suggestions would be useful.

carlocaione replied:
@ceolin you can still do that, see https://github.com/zephyrproject-rtos/zephyr/blob/main/drivers/mbox/mbox_handlers.c

carlocaione commented on tests/lib/devicetree/api/src/main.c:
```diff
@@ -2145,6 +2145,66 @@ static void test_mbox(void)
+	/* DT_INST_RESET_CELL */
+	zassert_equal(DT_INST_RESET_CELL(0, line), 10, "");
+
+	/* clock-freq on fixed clock */
```
`clock-freq`?

carlocaione commented on drivers/reset/Kconfig:
```diff
@@ -27,4 +27,6 @@ config RESET_INIT_PRIORITY
 
 comment "Reset Controller Drivers"
 
+source "drivers/reset/Kconfig.rpi_pico"
```
`rsource`?

carlocaione commented on include/devicetree/reset.h:
```diff
@@ -0,0 +1,246 @@
+ *         index "idx"
+ * @see DT_PHANDLE_BY_IDX()
+ */
+#define DT_RESET_CTLR_BY_IDX(node_id, idx) \
+	DT_PHANDLE_BY_IDX(node_id, resets, idx)
```
Usually we do not pollute the `DT_*` namespace, so you want to rename these functions something like `RESET_DT_...` instead.

andrei-edward-popa replied:
So DT_RESET_CTLR_BY_IDX need to be remaned as RESET_CTRL_DT_BY_IDX?

carlocaione replied:
`RESET_DT_CTLR_BY_IDX`

carlocaione commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,188 @@
+ * @param inst Instance number.
+ * @param id Reset line.
+ */
+#define RESET_DEV_CONFIG_INIT(inst, id)				\
```
better `RESET_DT_...`

carlocaione commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -302,6 +309,7 @@ static const struct uart_driver_api uart_rpi_driver_api = {
 		.uart_dev = uart##idx,						\
 		.uart_regs = (uart_hw_t *)uart##idx,				\
 		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),			\
+		.reset = RESET_DT_CONFIG_INIT(idx, id),				\
```
to keep consistency with the other subsystems, use `RESET_DT_INST_DEV_CONFIG_GET`.

Besides that LGTM.

gmarull reviewed (changes requested):
Thanks for this addition, I think it could also be useful for other platforms like STM32/GD32. However, in those platforms, the reset controller registers are also part of the same block used by clock controller, so we should see how it would work. Some documentation (Sphinx side) would also be good.

carlocaione commented on include/devicetree/reset.h:
```diff
@@ -0,0 +1,246 @@
+ *         index "idx"
+ * @see DT_PHANDLE_BY_IDX()
+ */
+#define RESET_DT_CTLR_BY_IDX(node_id, idx) \
```
uhm, quite the opposite I think. To not pollute the `DT_` namespace we should use a different prefix.

andrei-edward-popa replied:
@carlocaione What about include/devicetree/pinctrl.h? All pinctrl DT apis begin with DT_.

carlocaione replied:
I'm looking at `include/drivers/pinctrl.h` because this is a driver.

carlocaione replied:
Ooook I see the problem now. These macros should be moved to `include/drivers/reset.h`

gmarull replied:
There are 2 sides: the DT API part (include/devicetree/*) and the driver (include/drivers/*). AFAIK, the first needs needs `DT_`, the driver part `DRIVERNAME_`.

carlocaione replied:
right, but any good reason to not have everything in `include/drivers` like it was done for example for the `mbox` driver class?

andrei-edward-popa replied:
Because we want to create tests with this macros in tests/lib/devicetree/api maybe?

carlocaione replied:
you can still do that. Probably @mbolivar-nordic should have the final word here :)

mbolivar-nordic replied:
The current structure is OK if we want to have a DT API portion of this, but I am not sure if we do or not yet.

I have a few questions:

- same question as @gmarull about how this will work on platforms like STM32 that have a combined RCC (reset and clock control) peripheral; cc @erwango 
- is there a precedent for this sort of binding anywhere else? If we are working with a common paradigm inherited from Linux or something, I'll have fewer concerns, but if we are going off and inventing something new on our own, I'd want to know more about how this will generalize before extending the (stable) DT API

andrei-edward-popa replied:
> The current structure is OK if we want to have a DT API portion of this, but I am not sure if we do or not yet.
> 
> I have a few questions:
> 
> * same question as @gmarull about how this will work on platforms like STM32 that have a combined RCC (reset and clock control) peripheral; cc @erwango
> * is there a precedent for this sort of binding anywhere else? If we are working with a common paradigm inherited from Linux or something, I'll have fewer concerns, but if we are going off and inventing something new on our own, I'd want to know more about how this will generalize before extending the (stable) DT API

Okay, I will try to respond to that:

1)    I understand that some platforms a combined reset and clock control. I'm not really sure about how this works right now, the reset control part is used in some way with the clock API or only the clock control part is used? We can try to create 2 separate nodes in dts, one for clock control and the other one for reset control, both have the same reg address and use them separately. I'm not an expert, so I would gladly receive any suggestions for devices with RCC.
2)    Those bindings are more or less inherited from Linux. Linux uses a reset line 'unsigned long id' for information about resetting a device and this information can be passed directly into 'resets' binding for different devices. So I think that we didn't invent something new for this.

About DT API reset portion, I will modify it in the way that will be established.

mbolivar-nordic replied:
> Those bindings are more or less inherited from Linux. Linux uses a reset line 'unsigned long id' for information about resetting a device and this information can be passed directly into 'resets' binding for different devices. So I think that we didn't invent something new for this.

Can you give me a few pointers to how this works in the Linux source code?

andrei-edward-popa replied:
Yes, sure.
https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/reset/xlnx,zynqmp-reset.txt
https://elixir.bootlin.com/linux/latest/source/drivers/reset/reset-zynqmp.c
https://elixir.bootlin.com/linux/latest/source/arch/arm64/boot/dts/xilinx/zynqmp.dtsi#L816

mbolivar-nordic replied:
Thank you @andrei-edward-popa; this is very helpful.

I also see the following:

- https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/reset/reset.txt documents the general expectations for reset providers and consumers. This is consistent with the new API in include/devicetree/reset.h in this PR. I agree that if we are going to support bindings with `#reset-cells` and `resets` properties, we should have this API in devicetree.h, for consistency with other things like clocks, dmas, etc.
- the stm32 binding in https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/reset/st,stm32-rcc.txt is basically just a link to https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/clock/st,stm32-rcc.txt, which just documents that `#reset-cells` should be 1 for this IP, and documents [here](https://elixir.bootlin.com/linux/latest/source/Documentation/devicetree/bindings/clock/st,stm32-rcc.txt#L121) how to use this properly to set the 1 cell needed to identify a reset line on those SoCs from a consumer node
- https://elixir.bootlin.com/linux/latest/source/include/linux/reset-controller.h looks like it contains the driver API

Overall, the namespacing and organization of the DT macros look good to me; consider this message my +1 on the approach from the devicetree side.

The existence of a corresponding driver API in zephyr also sounds like it could be useful, but I haven't thought about it much. The design and implementation of the higher-level driver APIs I will leave to you all to figure out :).

carlocaione replied:
> Overall, the namespacing and organization of the DT macros look good to me; consider this message my +1 on the approach from the devicetree side.

Is it though? If we want to keep the DT macros in `include/devicetree/reset.h` then I think the namespace should be `DT_RESET_*` instead of `RESET_DT_*` to be consistent with what we do with the other subsystems.

andrei-edward-popa replied:
@carlocaione Check the commits, right now it is the version with DT_RESET.

carlocaione commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,188 @@
+ *
+ * @return 0 on success, negative code on error.
+ */
+__syscall int reset_reset(const struct reset_dev_config *reset);
```
`reset_toggle`?

andrei-edward-popa replied:
Should I change also the name from reset_driver_api from 'reset_api_reset reset' to 'reset_api_toggle toggle'?

gmarull commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,192 @@
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
+
+	if (!api) {
+		return -ENOTSUP;
```
`-ENOSYS`

andrei-edward-popa replied:
Can I add also some checks like if (!api->assert) { return -ENOSYS; }?

gmarull replied:
Sorry, I realized now that this is de-referencing twice. I think we should do then:

1. An API that just takes a `const struct device *` + `uint32_t id`
2. A "_dt" API that takes a sort of `_dt_spec` config (see for example GPIO or I2C APIs, they use this pattern)

There's no need to check for NULL API, I think. I misunderstood the code and thought that the API was allowed to not implement the call.

gmarull replied:
please check for the `_dt` approach, as other APIs like GPIO/I2C/SPI do.

https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/i2c.h#L67
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/i2c.h#L82
https://github.com/zephyrproject-rtos/zephyr/blob/main/include/drivers/i2c.h#L609

gmarull commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,192 @@
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
+
+	if (!api) {
+		return -ENOTSUP;
```
`-ENOSYS`

gmarull commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,192 @@
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
+
+	if (!api) {
+		return -ENOTSUP;
```
`-ENOSYS`

gmarull commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,192 @@
+ *
+ * @param reset Reset controller device configuration.
+ *
+ * @return 0 on success, negative code on error.
```
I'd suggest to use `@retval`, since some error codes like `-ENOSYS` can be explained better:

```
@retval 0 On success
@retval -ENOSYS If the functionality is not implemented by the device
@retval -errno Other negative errno in case of failure
```

gmarull commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,192 @@
+
+static inline int z_impl_reset_toggle(const struct reset_dev_config *reset)
+{
+	const struct reset_driver_api *api = (const struct reset_driver_api *)reset->dev->api;
```
nit: remove redundant cast here (and others)

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -302,6 +309,7 @@ static const struct uart_driver_api uart_rpi_driver_api = {
 		.uart_dev = uart##idx,						\
 		.uart_regs = (uart_hw_t *)uart##idx,				\
 		.pcfg = PINCTRL_DT_INST_DEV_CONFIG_GET(idx),			\
+		.reset = RESET_DT_INST_DEV_CONFIG_INIT(idx, id),		\
```
this driver has a dependency on `CONFIG_RESET=y` now, maybe we should specify this in Kconfig. Same can be said on pinctrl and I think most drivers don't do that, though.

gmarull reviewed (commented):
Please have a look at my comments. A couple of points:

1. Maybe we should use `reset_controller_` prefix (as in Linux)?
2. Please add some `doc/` documentation. A page with a brief introduction, use cases, perhaps implementation guidelines (e.g., looks like some API calls are optional, which ones and why?). It doesn't need to be extensive.

The rest LGTM, I will try to draft a GD32/STM32 driver for this API.

andrei-edward-popa commented:
> Please have a look at my comments. A couple of points:
> 
>     1. Maybe we should use `reset_controller_` prefix (as in Linux)?
> 
>     2. Please add some `doc/` documentation. A page with a brief introduction, use cases, perhaps implementation guidelines (e.g., looks like some API calls are optional, which ones and why?). It doesn't need to be extensive.
> 
> 
> The rest LGTM, I will try to draft a GD32/STM32 driver for this API.

Ok, let's see:
1)    So you mean to rename for example reset_driver_api to reset_controller_driver_api?
2)    Where do I need to add this documentation? Under doc/reference/peripherals/reset.rst?

gmarull commented:
> @gmarull Ok, let's see:
> 
> 1. So you mean to rename for example reset_driver_api to reset_controller_driver_api?

Yes, what do others think @carlocaione ?

> 2. Where do I need to add this documentation? Under doc/reference/peripherals/reset.rst?

Yes, sounds like the right place

carlocaione commented:
> Yes, what do others think @carlocaione ?

Uhm, I think that we could just consider that as given to be honest (the use of `_controller_` doesn't seem very popular among driver classes looking into `include/drivers`)

gmarull commented:
> > Yes, what do others think @carlocaione ?
> 
> Uhm, I think that we could just consider that as given to be honest (the use of `_controller_` doesn't seem very popular among driver classes looking into `include/drivers`)

We have a hybrid: `clock_control` (maybe it should either be `clock` or `clock_controller` btw...)

gmarull commented on doc/reference/peripherals/reset.rst:
```diff
@@ -0,0 +1,23 @@
+Overview
+********
+
+Reset controllers are units that control the reset signals to multiple peripherals. The reset controller API allows peripheral drivers to request control over their reset input signals, including the ability to assert, deassert and toggle those signals. Also, the reset status of the reset input signal can be checked.
+
+Mainly, the assert and deassert API functions are optional because in most cases we want to toggle the reset signals.
```
nit: please wrap to 80cols as most docs do

gmarull replied:
add the API in https://docs.zephyrproject.org/latest/reference/api/overview.html

andrei-edward-popa replied:
@gmarull Experimental?

gmarull replied:
Yep, sounds like it should be experimental since there'll be just 1 implementation at the beginning, so the API may need adjustments as more are added.

andrei-edward-popa replied:
Okay, cool.

gmarull commented on doc/reference/api/overview.rst:
```diff
@@ -256,6 +256,11 @@ current :ref:`stability level <api_lifecycle>`.
+     
+   * - :ref:`reset_api`
+     - Experimental
+     - 3.0
+     - 3.0
```
should be 3.1, since it will be the first release with this API

andrei-edward-popa replied:
Done!

carlocaione reviewed (commented):
Can you update `doc/reference/devicetree/api.rst`?

gmarull reviewed (changes requested):
I think that to be in line with other APIs, like e.g., GPIO or I2C, we should be using the `_dt_spec` convention here as well for consistency. This means, for example:

```c
struct reset_dt_spec {
	/** Reset controller device. */
	const struct device *dev;
	/** Reset line. */
	uint32_t id;
};

/* provide helper macros like: */
#define RESET_DT_SPEC_GET()
#define RESET_DT_SPEC_INST_GET()

/* non-dt spec variant */
__syscall int reset_toggle(const struct device *dev, uint32_t id);

static inline int z_impl_reset_toggle(const struct device *dev, uint32_t id)
{
	const struct reset_driver_api *api = dev->api;

	if (!api->toggle) {
		return -ENOSYS;
	}

	return api->toggle(dev, id);
}

/* dt spec wrapper */
static inline int reset_toggle_dt(const struct reset_dt_spec *spec)
{
	return reset_toggle(spec->dev, spec->id);
}
```

andrei-edward-popa commented:
> Can you update `doc/reference/devicetree/api.rst`?

Done!

andrei-edward-popa commented:
> I think that to be in line with other APIs, like e.g., GPIO or I2C, we should be using the `_dt_spec` convention here as well for consistency. This means, for example:
> 
> ```c
> struct reset_dt_spec {
> 	/** Reset controller device. */
> 	const struct device *dev;
> 	/** Reset line. */
> 	uint32_t id;
> };
> 
> /* provide helper macros like: */
> #define RESET_DT_SPEC_GET()
> #define RESET_DT_SPEC_INST_GET()
> 
> /* non-dt spec variant */
> __syscall int reset_toggle(const struct device *dev, uint32_t id);
> 
> static inline int z_impl_reset_toggle(const struct device *dev, uint32_t id)
> {
> 	const struct reset_driver_api *api = dev->api;
> 
> 	if (!api->toggle) {
> 		return -ENOSYS;
> 	}
> 
> 	return api->toggle(dev, id);
> }
> 
> /* dt spec wrapper */
> static inline int reset_toggle_dt(const struct reset_dt_spec *spec)
> {
> 	return reset_toggle(spec->dev, spec->id);
> }
> ```

I added all of this.

gmarull commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,340 @@
+{
+	const struct reset_driver_api *api = dev->api;
+
+	if (!api) {
+		return -ENOSYS;
+	}
```
nit: this can be removed, it's a programming error to implement a reset driver without API.

andrei-edward-popa replied:
Ok, I will remove it.

gmarull commented on drivers/serial/uart_rpi_pico.c:
```diff
@@ -68,9 +70,14 @@ static int uart_rpi_init(const struct device *dev)
 
+	/*
+	 * Can be used for resetting the UART device.
+	 * reset_toggle_dt(&config->reset);
```
note: why is this commented out?

andrei-edward-popa replied:
Because the uart_init function from Pico SDK resets the uart device, after the implementation of a clock controller driver for Pico we can remove the uart_init call and use our clock and reset controller for this.

gmarull replied:
if the driver is not using reset controller now, better remove it until it's possible.

andrei-edward-popa replied:
Okay, I removed it.

gmarull reviewed (dismissed):
LGTM, great work!

mbolivar-nordic commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,325 @@
+/**
+ * @brief Put the device in reset state
+ *
+ * This function sets/cleares the reset bits of the device,
```
Please fix the typo: "clears", not "cleares", here and elsewhere in this patch.

andrei-edward-popa replied:
Done!

mbolivar-nordic commented on include/devicetree/reset.h:
```diff
@@ -0,0 +1,309 @@
+ * @return the cell value at index "idx"
+ * @see DT_RESET_CELL_BY_IDX()
+ */
+#define DT_RESET_INST_CELL_BY_IDX(inst, idx, cell) \
```
The rule we follow in our API names is that `DT_INST_FOO(inst, ...)` is the same as `DT_FOO(DT_DRV_INST(inst), ...)`. These macro implementations look OK to me, but the names are not following the pattern.

For example, this should be `DT_INST_RESET_CELL_BY_IDX`, not `DT_RESET_INST_CELL_BY_IDX`.

Please change all of the `DT_RESET_INST_FOO` macros to `DT_INST_RESET_FOO` to follow the rule.

andrei-edward-popa replied:
I see, I modified this.

mbolivar-nordic commented on drivers/reset/Kconfig:
```diff
@@ -0,0 +1,30 @@
+# Reset Controller driver configuration options
+
+# Copyright (c) 2022 Andrei-Edward Popa <andrei.popa105@yahoo.com>
```
Copyright email does not match git signed-off-by email. Is this intentional?

mbolivar-nordic replied:
Note that this is just a question; I'm not requesting a change for this item.

andrei-edward-popa replied:
Thanks for pointing that out, I forgot to change the email address for the git commits with my personal email address, now it is fixed.

andrei-edward-popa replied:
@mbolivar-nordic The documentation is failing because of this:

 Warning, treated as error:
/home/runner/work/zephyr/zephyr/doc/_build/src/releases/release-notes-3.0.rst:655:Unknown interpreted text role "kconfig".

Probably because at line 655 needs to be kconfig:option?

carlocaione replied:
rebased

andrei-edward-popa replied:
Cool! 😊

mbolivar-nordic reviewed (changes requested):
I am in favor of merging this patch's changes to the devicetree API. I just have a few small issues to point out before I will approve.

@andrei-edward-popa this is nice work; well done.

andrei-edward-popa committed:
dts: bindings: added bindings for Reset Controller Driver

added bindings and compatible for reset controller nodes
added bindings for devices that use the reset controller

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
doc: reference: added documentation for Reset Controller

added documentation for Reset Controller Driver

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
dts: bindings: reset: added bindings for Raspberry Pi Pico

added reset bindings for Raspberry Pi Pico board

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
dts: bindings: serial: added reset bindings for Raspberry Pi Pico

added reset controller bindings for Raspberry Pi Pico board

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
include: drivers: added header file for Reset Controller Driver

added API and syscalls for reset controller
added reset controller devicetree macro public API header file

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
drivers: serial: rpi_pico: added reset support for UART devices

added reset controller for Raspberry Pi Pico UART devices

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
dts: arm: rpi_pico: added resets for UART instances

added resets for Raspberry Pi Pico UART instances
enabled reset controller in defconfig

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
tests: lib: devicetree: api: added test for reset public API

added test reset for reset public API macros

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
drivers: reset: added Reset Controller for Raspberry Pi Pico

added Reset Controller driver for Raspberry Pi Pico board

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
drivers: added file structure for Reset Controller Driver

added needed files for reset controller driver

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
dts: arm: rpi_pico: added reset controller node

added reset controller node in dtsi
added defines for all reset pins

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

andrei-edward-popa committed:
dts: bindings: test: added bindings for devicetree test reset

added bindings for vnd,reset used for devicetree test reset
added adc temp sensor as supporting reset

Signed-off-by: Andrei-Edward Popa <andrei.popa105@yahoo.com>

carlescufi commented:
@mbolivar-nordic can you please take a look?

carlescufi commented on include/drivers/reset.h:
```diff
@@ -0,0 +1,325 @@
+ *
+ * @param dev Reset controller device.
+ * @param id Reset line.
+ * @param status Where to write the reset status.
```
@andrei-edward-popa what is the format of this "status"? I see there's no enum or bool, just a `uint8_t`, so it'd be good to document what 'status' means in this context? At least a comment in the PR.

andrei-edward-popa replied:
Status is the current status of the reset, 0 if reset is deasserted and 1 if reset is asserted. This is only 1 bit of the status, the unused 7 bits can be used for other information about the reset or not used at all. Let me know if you think that we need to use bool instead of uint8_t.

carlescufi replied:
No, this sounds good. Thanks for the clarification.

mbolivar-nordic commented:
Approved and rebased one more time to try to get green CI.
